src/wormhole/cli/cmd_send.py

```diff
@@ -335,61 +337,41 @@ def _build_offer(self):
-            # zipfile in py3.7.0 that expects .seekable
-            if not hasattr(fd_to_send, "seekable"):
-                # AFAICT all the filetypes that STF wraps can seek
-                fd_to_send.seekable = lambda: True
-            num_files = 0
-            num_bytes = 0
-            tostrip = len(what.split(os.sep))
-            with zipfile.ZipFile(
-                    fd_to_send,
-                    "w",
-                    compression=zipfile.ZIP_DEFLATED,
-                    allowZip64=True) as zf:
-                for path, dirs, files in os.walk(what):
-                    # path always starts with args.what, then sometimes might
-                    # have "/subdir" appended. We want the zipfile to contain
-                    # "" or "subdir"
-                    localpath = list(path.split(os.sep)[tostrip:])
-                    for fn in files:
-                        archivename = os.path.join(*tuple(localpath + [fn]))
-                        localfilename = os.path.join(path, fn)
-                        try:
-                            zf.write(localfilename, archivename)
-                            num_bytes += os.stat(localfilename).st_size
-                            num_files += 1
-                        except OSError as e:
-                            errmsg = u"{}: {}".format(fn, e.strerror)
-                            if self._args.ignore_unsendable_files:
-                                print(
-                                    u"{} (ignoring error)".format(errmsg),
-                                    file=args.stderr)
-                            else:
-                                raise UnsendableFileError(errmsg)
-            fd_to_send.seek(0, 2)
-            filesize = fd_to_send.tell()
-            fd_to_send.seek(0, 0)
+            # We're sending a directory, stream it as a zipfile
+
+            zs = ZipStream(sized=True)
+            for filepath in walk(what, preserve_empty=True, followlinks=True):
+                try:
+                    if not os.access(filepath, os.R_OK):
+                        raise PermissionError(errno.EACCES, os.strerror(errno.EACCES), filepath)
+                    zs.add_path(
+                        filepath,
+                        arcname=os.path.relpath(filepath, what),
+                        recurse=False,
+                    )
+                except OSError as e:
+                    errmsg = "{}: {}".format(filepath, e.strerror)
+                    if not self._args.ignore_unsendable_files:
+                        raise UnsendableFileError(errmsg)
+                    print(
+                        "{} (ignoring error)".format(errmsg),
+                        file=args.stderr
+                    )
+
+            filesizes = [x["size"] for x in zs.info_list() if not x["is_dir"]]
+            filesize = len(zs)
             offer["directory"] = {
                 "mode": "zipfile/deflated",
```

Comment:
Note: I left this even though it's not technically correct anymore. `"zipfile/stored"` would be more accurate since compression is no longer used. I left it because changing it would break compatibility with existing clients that don't know how to handle `"zipfile/stored"` and it doesn't really matter if compression was used while extracting it. Open to alternative suggestions.

Reply:
That sounds like the right choice to me.

Reply:
So I checked what the other clients currently do. Wormhole William does not check the value at all, it just assumes it is a zip file and unzips it. wormhole-rs doesn't check either, but it doesn't need to because it passes on the compressed file without decompressing it for the user, so any format would be "fine". The only implementation which would not accept it is the Python code, which actually does check the value.

I do think that `zipfile/deflated` is very unfortunate to begin with for the reasons you mention, but also that `zipfile/stored` wouldn't be any better for the same reason. Ideally, the value should be simply `zipfile`. Of course this is too late to change now and therefore using `zipfile/deflated` is the best thing to do currently. But I think at the very least that the check in the Python version should be softened to also accepting `zipfile`, so that we may hopefully be able to migrate some time down in the future.

https://github.com/magic-wormhole/magic-wormhole/blob/master/src/wormhole/cli/cmd_receive.py#L332

Reply:
@piegamesde implemented in the latest commit